src/v/cloud_storage_clients/client_pool.cc

```diff
@@ -463,9 +482,10 @@ void client_pool::release(http_client_ptr leased) {
       "releasing a client, pool size: {}, capacity: {}",
       _pool.size(),
       _capacity);
-    if (_pool.size() < _capacity) {
-        _pool.emplace_back(std::move(leased));
-    }
+    vassert(
+      _pool.size() < _capacity,
+      "tried to release a client but the pool is at capacity");
+    _pool.emplace_back(std::move(leased));
```

Comment:
that's a pretty big change of behavior, is this why the test was hanging?

Reply:
It is related to this, yes. The change in behavior doesn’t have a material effect. Used ManyPartitionsTest with tiered storage to confirm that.